fluka parsing geometry figures and zones

Bogdan2423 opened the pull request.

Bogdan2423 commented:
I think this is ready for review.
I tested it manually with colimator geometry:
![obraz](https://github.com/yaptide/converter/assets/92746092/111ea199-f9c5-42cb-9bfe-bd0ba4aff8ce)
The result looked like this:
![25](https://github.com/yaptide/converter/assets/92746092/7a84dba9-0ace-448b-8828-309ab582d0d0)
Not sure how to evaluate this.
Here is the input file that was generated:
[fl_sim.txt](https://github.com/yaptide/converter/files/13238083/fl_sim.txt)
I set aspects other than geometry (beam, scorers) manually, so there may be errors there

grzanka commented on converter/common.py:
```diff
@@ -36,3 +37,56 @@ def get_configs_json(self) -> dict:
+    # abs() to fix that, but we need to remember the sign and update `n` accordingly
+    sign = 1
+
+    if result < 0:
+        result = abs(result)
```
```suggestion
    if number < 0:
        result = abs(number)
```

better to work on original values

grzanka commented on converter/fluka/input.py:
```diff
@@ -1,14 +1,20 @@
-    """Class mapping of the fluka input file."""
+    """Class mapping of the Fluka input file."""
 
     energy_GeV: float = 0.15  # GeV FLUKA specific
     number_of_particles: int = 10000
```
```suggestion
    energy_GeV: float = 0.07  # GeV FLUKA specific
    number_of_particles: int = 10000
```

Lets take change hardcoded value to 70 MeV, to be more consistent with documentation https://yaptide.github.io/docs/settings/

grzanka reviewed (changes requested):
small comments

grzanka commented:
> I think this is ready for review. I tested it manually with colimator geometry: ![obraz](https://user-images.githubusercontent.com/92746092/279956281-111ea199-f9c5-42cb-9bfe-bd0ba4aff8ce.png) The result looked like this: ![25](https://user-images.githubusercontent.com/92746092/279956380-7a84dba9-0ace-448b-8828-309ab582d0d0.png) Not sure how to evaluate this. Here is the input file that was generated: [fl_sim.txt](https://github.com/yaptide/converter/files/13238083/fl_sim.txt) I set aspects other than geometry (beam, scorers) manually, so there may be errors there

I really like your test with collimator. Something seems strange in that example. Some of the protons with initial energy of 70 MeV should stop in the collimator material.

Cross-check the BEAM card (https://flukafiles.web.cern.ch/manual/chapters/description_input/description_options/beam.html#beam)

![image](https://github.com/yaptide/converter/assets/7374733/b194dc12-e7b8-4955-9ac6-4488970800f9)

We deal with kinetic energy in MeV, not with momentum. Therefore you should write somewhere there -0.07 if you want protons with 70 MeV kinetic energy. 

As for energy spread you can put zero.

Bogdan2423 commented:
> > I think this is ready for review. I tested it manually with colimator geometry: ![obraz](https://user-images.githubusercontent.com/92746092/279956281-111ea199-f9c5-42cb-9bfe-bd0ba4aff8ce.png) The result looked like this: ![25](https://user-images.githubusercontent.com/92746092/279956380-7a84dba9-0ace-448b-8828-309ab582d0d0.png) Not sure how to evaluate this. Here is the input file that was generated: [fl_sim.txt](https://github.com/yaptide/converter/files/13238083/fl_sim.txt) I set aspects other than geometry (beam, scorers) manually, so there may be errors there
> 
> I really like your test with collimator. Something seems strange in that example. Some of the protons with initial energy of 70 MeV should stop in the collimator material.
> 
> Cross-check the BEAM card (https://flukafiles.web.cern.ch/manual/chapters/description_input/description_options/beam.html#beam)
> 
> ![image](https://user-images.githubusercontent.com/7374733/280011087-b194dc12-e7b8-4955-9ac6-4488970800f9.png)
> 
> We deal with kinetic energy in MeV, not with momentum. Therefore you should write somewhere there -0.07 if you want protons with 70 MeV kinetic energy.
> 
> As for energy spread you can put zero.

Changed the BEAM definition to `BEAM            -0.07      0.0       0.0       3.0       0.0     -1.0 PROTON`
Now the result looks like this:
![25](https://github.com/yaptide/converter/assets/92746092/8e0c9fba-e3c5-451b-9d8e-2b32818be2e5)

grzanka commented on converter/fluka/cards/figure_card.py:
```diff
@@ -0,0 +1,47 @@
+                         f" {format_float(figure.y_max, n=16):+#}"
+                         f" {format_float(figure.z_min, n=16):+#}"
+                         f" {format_float(figure.z_max, n=16):+#}")
+            elif type(figure) is FlukaCylinder:
+                line += (f" {format_float(figure.coordinates[0], n=16):+#}"
+                         f" {format_float(figure.coordinates[1], n=16):+#}"
+                         f" {format_float(figure.coordinates[2], n=16):+#}"
+                         f" {format_float(figure.height_vector[0], n=16):+#}"
+                         f" {format_float(figure.height_vector[1], n=16):+#}\n"
+                         f"{format_float(figure.height_vector[2], n=16):+#}"
+                         f" {format_float(figure.radius, n=16):+#}")
```
Can you add here a line of comment as well, so instead of:
```
RCC fig2 +0.0 +0.0 -0.5 +0.0 +0.0
+1.0 +4.0
```

we should have

```
* cylinder fig2, bottom center (0,0,-0.5), spanning vector (0,0,1), 
* radius 4, height XX cm
* rotation angles: X*, Y*, Z*
RCC fig2 +0.0 +0.0 -0.5 +0.0 +0.0
+1.0 +4.0
```

information may be redundant, but this way we increase readability. 
Same for other figures.

Bogdan2423 replied:
ok, I made these changes

grzanka commented:
Looking at your config I see several issues:
- you place beam at Z = -1.5 cm, in a `boundary` region. Beam source should be in a region with normal material
- you have `world` material filled with water. That is unnecessary. Take a look a documentation in https://yaptide.github.io/docs/geometry/#main-box . There I set the world volume to be made of air.

Bogdan2423 commented:
> Looking at your config I see several issues:
> 
>     * you place beam at Z = -1.5 cm, in a `boundary` region. Beam source should be in a region with normal material

I was following documentation, [here ](https://yaptide.github.io/docs/settings/)it is placed at -1,5cm, is it a mistake or am I misunderstanding something? Anyway I moved the beam to Z= -1.0 cm

>     * you have `world` material filled with water. That is unnecessary. Take a look a documentation in https://yaptide.github.io/docs/geometry/#main-box . There I set the world volume to be made of air.
I changed the material to air.
After these two changes this is the result:
![25](https://github.com/yaptide/converter/assets/92746092/b648e250-5b22-4f08-9dc9-4f446b11471b)
Looks pretty much the same to me

grzanka commented:
> > Looking at your config I see several issues:
> > ```
> > * you place beam at Z = -1.5 cm, in a `boundary` region. Beam source should be in a region with normal material
> > ```
> 
> I was following documentation, [here ](https://yaptide.github.io/docs/settings/)it is placed at -1,5cm, is it a mistake or am I misunderstanding something? Anyway I moved the beam to Z= -1.0 cm

I was more concerned about the fact in which volume the source was, rather than where exactly is located on Z axis.

> > ```
> > * you have `world` material filled with water. That is unnecessary. Take a look a documentation in https://yaptide.github.io/docs/geometry/#main-box . There I set the world volume to be made of air.
> > ```
> 
> I changed the material to air. After these two changes this is the result: ![25](https://user-images.githubusercontent.com/92746092/280111228-b648e250-5b22-4f08-9dc9-4f446b11471b.png) Looks pretty much the same to me

Then another problem I see

```
world 5 +figworld -fig0 -fig1 -fig2 -fig3
```

I should be enough to specify
```
world 5 +figworld -fig0
```

Subtracting `fig1`, `fig2` and `fig3` which are already contained in `fig0` may lead to some strange side effects.

Can you reduce manually number of regions to the same as in SHIELD-HIT12A simulation ?

grzanka commented on converter/fluka/cards/figure_card.py:
```diff
@@ -18,27 +18,60 @@ def __str__(self) -> str:
+                y_max = format_float(figure.y_max, n=16)
+                z_min = format_float(figure.z_min, n=16)
+                z_max = format_float(figure.z_max, n=16)
+                line += (f"* box {figure.name}\n"
+                         f"* X range {x_min:+#}, {x_max:+#}\n"
+                         f"* Y range {y_min:+#}, {y_max:+#}\n"
+                         f"* Z range {z_min:+#}, {z_max:+#}\n"
```
Can you add here the same information which user types in the yaptide, namely X,Y,Z size lengths ?

Bogdan2423 replied:
ok

grzanka commented on tests/fluka/expected_fluka_output/fl_sim.py:
```diff
@@ -0,0 +1,88 @@
+RPP fig1 -4.0 +4.0 -4.0 +4.0 +0.0 +8.0
+* cylinder fig2
+* bottom center (+0.0, +0.0, -0.5), spanning vector (+0.0, +0.0, +1.0),
+* radius +4.0, height +0.0 cm
```
Why height is zero ?

Bogdan2423 replied:
Typo in variable name

Bogdan2423 commented:
> > > Looking at your config I see several issues:
> > > ```
> > > * you place beam at Z = -1.5 cm, in a `boundary` region. Beam source should be in a region with normal material
> > > ```
> > 
> > 
> > I was following documentation, [here ](https://yaptide.github.io/docs/settings/)it is placed at -1,5cm, is it a mistake or am I misunderstanding something? Anyway I moved the beam to Z= -1.0 cm
> 
> I was more concerned about the fact in which volume the source was, rather than where exactly is located on Z axis.
> 
> > > ```
> > > * you have `world` material filled with water. That is unnecessary. Take a look a documentation in https://yaptide.github.io/docs/geometry/#main-box . There I set the world volume to be made of air.
> > > ```
> > 
> > 
> > I changed the material to air. After these two changes this is the result: ![25](https://user-images.githubusercontent.com/92746092/280111228-b648e250-5b22-4f08-9dc9-4f446b11471b.png) Looks pretty much the same to me
> 
> Then another problem I see
> 
> ```
> world 5 +figworld -fig0 -fig1 -fig2 -fig3
> ```
> 
> I should be enough to specify
> 
> ```
> world 5 +figworld -fig0
> ```
> 
> Subtracting `fig1`, `fig2` and `fig3` which are already contained in `fig0` may lead to some strange side effects.
> 
> Can you reduce manually number of regions to the same as in SHIELD-HIT12A simulation ?

Result after removing the automatically generated "world" region:
![25](https://github.com/yaptide/converter/assets/92746092/ffcc1ad9-625a-4358-aa90-e704e9c3ff71)
Input file:
[fl_sim.txt](https://github.com/yaptide/converter/files/13249764/fl_sim.txt)
I also tried removing the surrounding black hole region, but after that the simulation crashes with the following error file:
[fl_sim001.txt](https://github.com/yaptide/converter/files/13249789/fl_sim001.txt)

grzanka commented:
> > > > Looking at your config I see several issues:
> > > > ```
> > > > * you place beam at Z = -1.5 cm, in a `boundary` region. Beam source should be in a region with normal material
> > > > ```
> > > 
> > > 
> > > I was following documentation, [here ](https://yaptide.github.io/docs/settings/)it is placed at -1,5cm, is it a mistake or am I misunderstanding something? Anyway I moved the beam to Z= -1.0 cm
> > 
> > 
> > I was more concerned about the fact in which volume the source was, rather than where exactly is located on Z axis.
> > > > ```
> > > > * you have `world` material filled with water. That is unnecessary. Take a look a documentation in https://yaptide.github.io/docs/geometry/#main-box . There I set the world volume to be made of air.
> > > > ```
> > > 
> > > 
> > > I changed the material to air. After these two changes this is the result: ![25](https://user-images.githubusercontent.com/92746092/280111228-b648e250-5b22-4f08-9dc9-4f446b11471b.png) Looks pretty much the same to me
> > 
> > 
> > Then another problem I see
> > ```
> > world 5 +figworld -fig0 -fig1 -fig2 -fig3
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I should be enough to specify
> > ```
> > world 5 +figworld -fig0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Subtracting `fig1`, `fig2` and `fig3` which are already contained in `fig0` may lead to some strange side effects.
> > Can you reduce manually number of regions to the same as in SHIELD-HIT12A simulation ?
> 
> Result after removing the automatically generated "world" region: ![25](https://user-images.githubusercontent.com/92746092/280267458-ffcc1ad9-625a-4358-aa90-e704e9c3ff71.png) Input file: [fl_sim.txt](https://github.com/yaptide/converter/files/13249764/fl_sim.txt) I also tried removing the surrounding black hole region, but after that the simulation crashes with the following error file: [fl_sim001.txt](https://github.com/yaptide/converter/files/13249789/fl_sim001.txt)

What about the beam source ? Are the proton emitted from a disc with radius 3cm as described here:
https://yaptide.github.io/docs/settings/ ?

grzanka commented:
I was looking at:

```
*...+....1....+....2....+....3....+....4....+....5....+....6....+....7....+....8
BEAM               07      0.0       0.0       3.0       0.0      -1.0PROTON
* beam source position
*...+....1....+....2....+....3....+....4....+....5....+....6....+....7....+....8
BEAMPOS          0.0       0.0      -1.0
```

there is something wrong with energy, it should be 0.07, not 07.

Also - check if BEAMPOS is correctly aligned (if using fixed format)

Bogdan2423 commented:
I think the beam energy was incorrectly aligned. After changing it these are the results for fluence and dose:
![25](https://github.com/yaptide/converter/assets/92746092/06b2ed8c-f648-4795-ab64-0d0b1db63777)
![26](https://github.com/yaptide/converter/assets/92746092/a26b1398-b02d-40d4-ba63-59fb5c2736b7)
Input file:
[fl_sim.txt](https://github.com/yaptide/converter/files/13253265/fl_sim.txt)
